arm9/source/utils/scripting.c

```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
+        bool al = !ww && off_disp_chars && (ptr != ptr_next);
+        bool ar = !ww && (llen_chars + 1 > off_disp_chars + TV_LLEN_DISP);
 
         // set text color / find start of comment of scripts
         u32 color_text = (nln == mno) ? script_color_active : (is_script) ? script_color_code : (u32) COLOR_TVTEXT;
-        int cmt_start = TV_LLEN_DISP; // start of comment in current displayed line (may be negative)
+        int cmt_start_bytes = TV_LLEN_DISP; // start of comment in current displayed line (may be negative)
+        int cmt_start_chars = 0;
         if (is_script && (nln != mno)) {
-            char* hash = line_seek(text, len, 0, ptr, 0);
-            for (; *hash != '#' && (hash - ptr < (int) llen); hash++);
-            cmt_start = (hash - ptr) - off_disp;
+            const char* hash = line_start(text, len, 0, ptr);
+            for (; *hash != '#' && hash - ptr < (int) llen_bytes; hash++);
+            cmt_start_bytes = hash - (ptr + off_disp_bytes);
+            if (cmt_start_bytes <= 0) color_text = script_color_comment;
+            else cmt_start_chars = chars_in_bytes(ptr + off_disp_bytes, cmt_start_bytes);
         }
-        if (cmt_start <= 0) color_text = script_color_comment;
 
         // build text string
-        snprintf(txtstr, sizeof(txtstr), "%-*.*s", (int) TV_LLEN_DISP, (int) TV_LLEN_DISP, "");
-        if (ncpy) memcpy(txtstr, ptr + off_disp, ncpy);
+        snprintf(txtstr, sizeof(txtstr), "%-*.*s", (int) (TV_LLEN_DISP * MAX_CHAR_SIZE), (int) (TV_LLEN_DISP * MAX_CHAR_SIZE), "");
+        if (ncpy_bytes) {
+            memcpy(txtstr, ptr + off_disp_bytes, ncpy_bytes);
+        }
         for (char* d = txtstr; *d; d++) if (*d < ' ') *d = ' ';
-        if (al) memcpy(txtstr + p_al, al_str, strlen(al_str));
-        if (ar) memcpy(txtstr + p_ar, ar_str, strlen(ar_str));
+        if (ar) {
+            char* textstr_end = txtstr + ncpy_bytes;
+            u32 txtstr_ar_bytes = 0;
+            for (int txtstr_ar_i = 0; txtstr_ar_i < (int) strlen(ar_str); txtstr_ar_bytes += GetPrevCharSize(textstr_end - txtstr_ar_bytes), ++txtstr_ar_i);
+            memcpy(textstr_end - txtstr_ar_bytes, ar_str, sizeof(ar_str));
+        }
+        if (al) {
+            u32 txtstr_al_bytes = bytes_in_chars_u32(txtstr, strlen(al_str));
+            memmove(txtstr + strlen(al_str), txtstr + txtstr_al_bytes, sizeof(txtstr) - txtstr_al_bytes);
+            memcpy(txtstr, al_str, strlen(al_str));
+        }
 
         // draw line number & text
         DrawString(TOP_SCREEN, txtstr, x_txt, y, color_text, COLOR_STD_BG);
         if (TV_LNOS > 0) { // line number
-            if (ptr != ptr_next)
-                DrawStringF(TOP_SCREEN, x_lno, y, ((ptr == text) || (*(ptr-1) == '\n')) ? COLOR_TVOFFS : COLOR_TVOFFSL, COLOR_STD_BG, "%0*lu", TV_LNOS, nln);
+            bool prev_ww_line_full = ww && ww == chars_between_pointers(line_seek_chars(text, len, ww, ptr, -1), ptr);
```

Comment:
Unfortunate edge case to draw line number when the `'\0'` is word-wrapped onto its own line...